### PR TITLE
#13: Implemented checkpoints after every epoch for the deepfake detector.

### PR DESCRIPTION
**Pull request by HueyAmiNhvtim**
## [What does this PR do?](https://google.github.io/eng-practices/review/developer/cl-descriptions)

This PR implemented checkpoints after every epoch during training so that if something bad happens, it can continue using the checkpoint from the last successful epoch. Also implemented restoration from old_models too, tho that one is
incomplete, and should be resolved later on.

## [Github Issue Number](https://docs.github.com/en/issues/tracking-your-work-with-issues/about-issues)

Resolves:

Relates to:
Nothing

## Author checklist
If any of these points have yet to be satisfied, make sure that you set the title to the format DRAFT #issue-num Title

- [x] The title is short and descriptive of the PR. Must start with the Github Issue Number (format: #issue-num Title).
- [x] The description follows proper cl description practices and mentions related Github Issues (make sure this is the first thing you mention).
- [x] Branch has merged in the **latest version of main**
- [x] **Linting** has occured, as per the project linting config 
- [x] All changed functions have proper **docstring** to describe what they do and how to use them.
 
 
## Reviewer checklist
Second reviewer will need to double check all of these points

- [ ] Relevant issue is mentioned in description 
- [ ] Code solves the issue
- [ ] Code is the best solution for the issue
- [ ] Branch is **ahead of main**
- [ ] Ensures the fix/feature works locally

### CODE REVIEW

File: models/visual_quality_discriminator.py [hunk 1]
```diff
@@ -202,24 +191,85 @@ def unison_shuffled_copies(a, b):
...
-model.compile(optimizer=tf.keras.optimizers.Adam(),
-              loss=tf.keras.losses.BinaryCrossentropy(), metrics=['accuracy', 'mean_absolute_error'])
-model.summary()
-
-model.fit(x_train, y_train, epochs=7)
-save_location = os.path.join(project_path, "saved_models/khoa")
-model.save(save_location)   # TODO: get dotenv working and make this an env variable
+
+
+def load_or_create_new_model():
```
Comment (BasicallyOk) [inline models/visual_quality_discriminator.py:196]: Bad idea, in a GAN setup, we do not want a premade model at all.
For the Synch Accuracy one we will implement that separately.

File: models/visual_quality_discriminator.py [hunk 2]
```diff
@@ -202,24 +191,85 @@ def unison_shuffled_copies(a, b):
...
+        print(f"Epoch {epoch} starting now:")
+        model.fit(x_train, y_train)
+        manager.save()
+
+    # delete all checkpoints after the entire run is complete.
+    for file in os.listdir(save_checkpoint_path):
+        os.remove(os.path.join(save_checkpoint_path, file))
+
+
+train()
```
Comment (BasicallyOk) [inline models/visual_quality_discriminator.py:272]: Put this in a if __name__ == "__main__": block.
We want this file to export the model, not train it

**Review by BasicallyOk — CHANGES_REQUESTED** [review]

File: models/visual_quality_discriminator.py (hunk 1 above)
Comment (BasicallyOk) [reply models/visual_quality_discriminator.py:196]: To resolve this, make sure that:

- Call the function load_model(), we won't need polymorphism, but it's good practice to be consistent and concise.
- If you want to load model, or train, do that in the main block. This file's main purpose is to define the model, the training and testing *must* be done only in a main block.

**Comment by BasicallyOk** [comment]
Make sure that all changes are mentioned in the description.
Only the checkpoint change was mentioned.

File: models/visual_quality_discriminator.py (hunk 1 above)
Comment (HueyAmiNhvtim) [reply models/visual_quality_discriminator.py:196]: So should we remove that model.save line in the visual quality discriminator model? Because the visual quality discriminator is not part of the GAN setup (as in not being trained alongside the sync_accuracy and the generator one) if I remember correctly.

File: models/visual_quality_discriminator.py (hunk 2 above)
Comment (HueyAmiNhvtim) [reply models/visual_quality_discriminator.py:272]: I suppose that I should also put the check point stuff not in the functions and the code for testing the model in that block then.

File: models/visual_quality_discriminator.py (hunk 1 above)
Comment (BasicallyOk) [reply models/visual_quality_discriminator.py:196]: Yes, it's still better to have consistent and predictable function design.

We'll just train and save it separately

File: models/visual_quality_discriminator.py (hunk 2 above)
Comment (BasicallyOk) [reply models/visual_quality_discriminator.py:272]: Yes, but which part of the checkpoints do you mean. We still wamt to have proper checkpoints implemented to avoid having to restart when anything happens
